### PR TITLE
Lazy load list of all Python Environments

### LINKED ISSUE #7583: Adopt new Python extensions non-blocking API for interpreter disocvery
Three discovery APIs are being added for Jupyter which is meant to replace existing `getInterpreters()` API: https://github.com/microsoft/vscode-python/pull/17452

```typescript
type PythonApiForJupyterExtension = {
     ...
    /**
     * IInterpreterService
     */
    readonly refreshPromise: Promise<void> | undefined;
    /**
     * IInterpreterService
     */
    readonly onDidChangeInterpreters: Event<PythonEnvironmentsChangedEvent>;
    /**
     * Equivalent to getInterpreters() in IInterpreterService
     */
    getKnownInterpreters(resource?: Uri): PythonEnvironment[];
    /**
     * Equivalent to getAllInterpreters() in IInterpreterService
     */
    getInterpreters(resource?: Uri): Promise<PythonEnvironment[]>;
    ...
```

### Previous behavior:

Existing `getInterpreters()` is an asynchronus blocking API, so it waits until all of discovery is completed the first time it is called. Subsequent calls return instantaneously from cache if the same resource is queried again.

### New behavior:

- New `getKnownInterpreters()` simply returns whatever is currently in cache. It is useful when we do not need all of the interpreters at once, which is usually the case. In the Python extension we are able to replace all instances of `getInterpreters()` with `getKnownInterpreters()`. Do note that the kind info returned by this API may change as refresh progresses.
- `onDidChangeInterpreters` fires when an environment gets added/removed/updated in cache. Event properties tracks the old and new environments accordingly.
- Python extension triggers a refresh in background when the extension loads for the first time, which is when `refreshPromise` is set. It can be used to implement now deprecated `getInterpreters()` in the following way:
```typescript
    /** Gets all interpreters **/
    public async getInterpreters(resource?: Uri): Promise<PythonEnvironment[]> {
        await this.refreshPromise; // Wait for the refresh to finish so that the cache contains all environments.
        return getKnownInterpreters(resource); // Return from cache.
    }
```
- `getInterpreterDetails()` is still available as before, and can be used to get complete and reliable information on environments. As it is a blocking call, it's recommend to only use it if complete information is needed, which is generally only the case for selected interpreters.

You can the see the new API in action here in our dynamic quickpick https://github.com/microsoft/vscode-python/pull/17043#issue-975156935.

> Note: Any internal interpreter list caching that was recently added should go away once `onDidChangeInterpreters` events are available.

### Changes regarding quickpick API

We also used to expose `getSuggestions` API for the quickpick, which has the same problem as `getInterpreters()`. The following non-blocking API is being added to replace it:
```typescript
getKnownSuggestions(resource: Resource): Promise<IInterpreterQuickPickItem[]>;
```
It can be used in combination of `refreshPromise` to get all suggestions. Let me know if you have any questions! 😄

### PR DESCRIPTION
**Pull request by DonJayamanne**
Fixes #7583

### CODE REVIEW

File: TELEMETRY.md [hunk 1]
```diff
@@ -1913,22 +1913,6 @@ Expand each section to see more information about that event.
         Common measurement used across a number of events.  
 
 
-* DS_INTERNAL.INTERPRETER_LISTING_PERF  (Telemetry.InterpreterListingPerf)
```
Comment (DonJayamanne) [inline TELEMETRY.md:1916]: This telemetry is no longer valid, as we no longer have an API to fetch all python enviornments.

File: src/test/datascience/execution.unit.test.ts [hunk 2]
```diff
@@ -1,1135 +0,0 @@
-// Copyright (c) Microsoft Corporation.
```
Comment (DonJayamanne) [inline src/test/datascience/execution.unit.test.ts:1]: See here https://github.com/microsoft/vscode-jupyter/pull/11673
This PR depends on Removed in separate PR https://github.com/microsoft/vscode-jupyter/pull/11673

File: src/kernels/raw/finder/localKernelFinder.node.ts [hunk 3]
```diff
@@ -90,38 +79,18 @@ export class LocalKernelFinder implements ILocalKernelFinder, IExtensionSyncActi
...
 
     private async loadInitialState() {
-        traceVerbose('LocalKernelFinder: load cache');
-        // loading cache, which is resource agnostic
-        const kernelsFromCache = await this.getFromCache();
-        let kernels: LocalKernelConnectionMetadata[] = [];
-
-        this.updateCache().catch(noop);
-
-        if (Array.isArray(kernelsFromCache) && kernelsFromCache.length > 0) {
```
Comment (DonJayamanne) [inline src/kernels/raw/finder/localKernelFinder.node.ts:117]: We need to cache anything anymore, seems pointless and complicated.
Can add this later if we need this.

File: src/platform/api/pythonApi.ts [hunk 4]
```diff
@@ -345,28 +355,23 @@ export class InterpreterService implements IInterpreterService {
...
+    public get environments(): readonly PythonEnvironmentV2[] {
+        this.getApi().catch(noop);
+        this.hookupOnDidChangeInterpreterEvent();
+        return this.api?.environments?.known || [];
+    }
 
-    @traceDecoratorVerbose('Get Interpreters', TraceOptions.Arguments | TraceOptions.BeforeCall)
-    public getInterpreters(): Promise<PythonEnvironment[]> {
-        const stopWatch = new StopWatch();
+    private getInterpreters(): Promise<PythonEnvironment[]> {
```
Comment (DonJayamanne) [inline src/platform/api/pythonApi.ts:369]: The method `getInterrpeters` is no longer used, we will not rely on a push model completely.

Comment (DonJayamanne) [reply src/platform/api/pythonApi.ts:369]: Fortunately this was used only in a couple of places in the extension code.

File: src/kernels/raw/finder/localKernelSpecFinderBase.node.ts [hunk 5]
```diff
@@ -72,11 +71,6 @@ export abstract class LocalKernelSpecFinderBase<
         disposables.push(this);
     }
 
-    @testOnlyMethod()
-    public clearCache() {
```
Comment (DonJayamanne) [inline src/kernels/raw/finder/localKernelSpecFinderBase.node.ts:76]: Unused

**Review by alexr00 — APPROVED** [review]

**Pull request merged**